File: leetcode/Python/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        p, c, sol, st, node, t = [[] for i in range(numCourses)], [[] for i in range(numCourses)], [], [], 0, 0
        for pre in prerequisites:
            p[pre[0]].append(pre[1])
            c[pre[1]].append(pre[0])
        for i in range(numCourses):
            if not p[i]:
                st.append(i)
        while st:
            node=st.pop()
            sol.append(node)
            for n in c[node]:
                t=p[n].index(node)
                p[n][t], p[n][-1] = p[n][-1], p[n][t]
                p[n].pop()
                if not p[n]:
                    st.append(n)
        return len(sol)==numCourses

File: leetcode/C++/course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> sol, v;
        vector<vector<int>> parent(numCourses), child(numCourses);
        vector<int>::iterator it;
        for(vector<int> &t : prerequisites){
            parent[t[0]].push_back(t[1]);
            child[t[1]].push_back(t[0]);
        }
        for(int i=0; i<numCourses; ++i) if(parent[i].empty()) v.push_back(i);
        while(!v.empty()){
            sol.push_back(v.back()); v.pop_back();
            for(int &t:child[sol.back()]){
                if((it=find(parent[t].begin(), parent[t].end(), sol.back())) != parent[t].end()){
                    *it=parent[t].back();
                    parent[t].pop_back();
                    if(parent[t].empty()) v.push_back(t);
                }
            }
        }
        return sol.size()==numCourses;
    }
};